02-intermediate-ruby/testing-with-minitest.md

```diff
@@ -0,0 +1,419 @@
+
+1. A `Bill` is constructed without an argument
+1. A `Bill` is constructed with a non-array argument
+1. Or Bill.new is called an empty list
+
+We will choose to **raise an error**.  By doing this we can insure that no `Bill` is created in an invalid state.
+
+So we will add another `it` block and use the `must_raise` expectation.
+
+```ruby
+it "will raise an error if created without a list of prices" do
+  proc {
+    bill = Bill.new
+  }.must_raise ArgumentError
+end
+```
+
+**Question**: The code above tests for the first situation.  How can you write a tests for the others?  You can find a solution [here](https://github.com/AdaGold/bill_calculator/commit/96f00a74e60d4e4246d0c8a10dfe66dea89cfd6d)
+
+#### How to make it pass?
+
+Now we need to write code to make the tests pass.  Raising an error is new for us right now, so lets look at how we can do so:
+
+```ruby
+# bill.rb
+
+class Bill
+  def initialize(prices)
+    if !(prices.is_a? Array) || prices.length < 1
+      raise ArgumentError.new("Bill requires a non-empty array")
+    end
+    @prices = prices
+  end
+end
+```
+
+The command `raise ArgumentError.new("...")` triggers an error, known in programming as an **exception** to occur which will rise up to the calling method and force the user of our class to either handle the error or prevent it from occuring.  In this case it prevents misuse of our class.  Anyone using our `Bill` must instantiate it with a non-empty list, or it will not work.  
+
+#### Question
+Consider other edge cases we might want?  What kinds of invalid arguments could get passed to initialize?
+
+#### Practice Exercise
+
+Now we need make a method that calculates the bill subtotal.  Write a test (`it` block) that creates an instance of Bill, calls a `subtotal` method and ensure that it returns the correct answer (the sum of the prices).  You can use the `must_equal` matcher.  
+
+Then make the test pass by editing `bill.rb`
+
+#### Check & Verify
+
+Check with your neighbor.  You can find a solution [here for the test](https://github.com/AdaGold/bill_calculator/commit/e5bbfe7cd0235e93408e6fae81abf5f2c8624437) [and here for `bill.rb`](https://github.com/AdaGold/bill_calculator/commit/c67fec9271b50fe1d2fd28cf93db343b2dee4f06)
+
+#### Arrange-Act-Assert
+
+Notice our example followed a pattern from our earlier [introduction to automated tests](../00-programming-fundamentals/intro-to-automated-tests.md).  First we created a `Bill`, or **arranged** the situation.  Then by calling the `subtotal` method we **acted**, or performed the action we wanted to test.  Lastly we used the expectation to **assert** that the result of our action was correct.    This is a good example of the _Arrange-Act-Assert_ pattern discussed earlier.  
+
+#### Practice Exercise - Sales Tax
```

Comment:
This feels like a lot of `h4`s (`####`) in a row. Maybe it's fine, but it might be there's an opportunity to add more structure here.